lib/sles4sap/console_redirection.pm

```diff
@@ -43,33 +43,133 @@ sub handle_login_prompt {
     }
     set_serial_term_prompt();    # set correct serial prompt
 }
 
 =head2 redirection_init
 
-    redirection_init();
+    redirection_init( [, ssh_user=>$ssh_user, target_ip=>$target_ip, ssh_tunnel_port=>$ssh_tunnel_port]);
+
+B<ssh_user>: Target login user - default value is defined by OpenQA parameter REDIRECT_TARGET_USER
+
+B<target_ip>: Target host IP - default value is defined by OpenQA parameter REDIRECT_TARGET_IP
+
+B<ssh_tunnel_port>: Local port used for reverse ssh tunnel, default: 22022
+
+Does initial setup for console redirection which includes:
+
+=over
+
+=item * sets BASE_VM_ID value with current machine id. This helps identifying the base (not redirected) VM.
 
-Do preparation before redirecting console. Gets base VM id and initial setup.
-This is required to be done only once at the beginning of the whole test.
-If you have a multi machine setup, execute this on each worker VM.
+=item * SSH key exchange for reverse ssh connection.
 
+=item * remote port forwarding for reverse SSH session
+
+=item * remote port forwarding of incoming traffic from remote VM to OpenQA server resources (upload_logs, download from 'data')
+
+=back
 =cut
 
 sub redirection_init {
+    my (%args) = @_;
+    $args{target_ip} //= get_required_var('REDIRECT_TARGET_IP');
+    $args{ssh_user} //= get_required_var('REDIRECT_TARGET_USER');
+    $args{ssh_tunnel_port} //= '22022';
+
+    croak 'Package autossh is not installed' if script_run('rpm -qi autossh');
+
+    record_info('Redirection init', 'Preparing console redirection');
+
     # This should get worker VM id before any redirection happening
     # ID serves as identification of the 'base' VM to return to.
-    set_var('WORKER_VM_ID', script_output 'cat /etc/machine-id');
+    set_var('BASE_VM_ID', script_output 'cat /etc/machine-id');
+
+    # Prepare keyless access from remote host to worker VM
+    connect_target_to_serial(%args);
+    script_run("sudo rm $reverse_ssh_key_base_name*", quiet => 1);
+    # remove any existing entry in known_hosts file
+    script_run("ssh-keygen -R [localhost]:$args{ssh_tunnel_port} -f ~/.ssh/known_hosts", quiet => 1);
+    script_run("sudo ssh-keygen -R [localhost]:$args{ssh_tunnel_port} -f /root/.ssh/known_hosts", quiet => 1);
+    assert_script_run("ssh-keygen -f $reverse_ssh_key_base_name -t rsa -b 2048 -N ''");
+    my $public_key = script_output("cat $reverse_ssh_key_base_name.pub", quiet => 1);
+    disconnect_target_from_serial(%args);
+
+    # Add cloud VM key to worker VM authorized keys
+    assert_script_run("echo \"$public_key\" >> ~/.ssh/authorized_keys", quiet => 1);
+
+    # Forward common ports.
+    # # Starts permanent reverse SSH connection from remote VM to worker VM.
+    remote_port_forward(destination_port => $args{ssh_tunnel_port},
+        ssh_user => $args{ssh_user},
+        destination_ip => 'localhost',
+        monitor_port => '20000');
```

Comment:
Port has to be unique as explained below. Hard-coding it here as `20000` and `20001` means only one tunnel and forwarding can exist at any time within a test. Right?

Reply:
Looks like it. When I was trying to run two forwards without monitor port, one would fail. 
Technically I should make it mandatory argument instead of disabled by default.

Reply:
> Technically I should make it mandatory argument instead of disabled by default.

why not? for self of simplicity?

Reply:
In case somebody calls functions 2x without specifying monitor port. 
I am not sure what will be more confusing. The failing test or to pick a port for monitoring.